packages/core/src/utils/zod.ts

```diff
@@ -1,20 +1,98 @@
 import { arbitraryObjectGuard } from '@logto/schemas';
-import { conditional } from '@silverhand/essentials';
+import { conditional, ValuesOf } from '@silverhand/essentials';
 import { OpenAPIV3 } from 'openapi-types';
 import {
   ZodArray,
   ZodBoolean,
+  ZodLiteral,
   ZodNativeEnum,
   ZodNullable,
   ZodNumber,
   ZodObject,
   ZodOptional,
   ZodString,
+  ZodStringDef,
+  ZodType,
+  ZodUnion,
   ZodUnknown,
 } from 'zod';
 
 import RequestError from '@/errors/RequestError';
 
+type ZodStringCheck = ValuesOf<ZodStringDef['checks']>;
+
+const zodStringCheckToSwaggerFormat = (zodStringCheck: ZodStringCheck) => {
+  const { kind } = zodStringCheck;
+
+  switch (kind) {
+    case 'email':
+    case 'url':
+    case 'uuid':
+    case 'cuid':
+    case 'regex':
+      return kind;
+
+    case 'min':
+    case 'max':
+      // Do nothing here
+      return;
+
+    default:
+      throw new RequestError('swagger.invalid_zod_type', zodStringCheck);
+  }
+};
+
+// https://github.com/colinhacks/zod#strings
+const zodStringToSwagger = (zodString: ZodString): OpenAPIV3.SchemaObject => {
+  const pattern = zodString._def.checks
+    .find((check): check is { kind: 'regex'; regex: RegExp } => check.kind === 'regex')
+    ?.regex.toString();
+
+  const minLength = zodString._def.checks.find(
+    (check): check is { kind: 'min'; value: number } => check.kind === 'min'
+  )?.value;
+  const maxLength = zodString._def.checks.find(
+    (check): check is { kind: 'max'; value: number } => check.kind === 'max'
+  )?.value;
+
+  const formats = zodString._def.checks
+    .map((zodStringCheck) => zodStringCheckToSwaggerFormat(zodStringCheck))
+    .filter((format) => format);
+
+  return {
+    type: 'string',
+    format: formats.length > 0 ? formats.join(' | ') : undefined,
+    minLength,
+    maxLength,
+    pattern,
+  };
+};
+
+// https://github.com/colinhacks/zod#literals
+const zodLiteralToSwagger = (zodLiteral: ZodLiteral<unknown>): OpenAPIV3.SchemaObject => {
+  const { value } = zodLiteral;
+
+  switch (typeof value) {
+    case 'boolean':
+      return {
+        type: 'boolean',
+        format: String(value),
+      };
+    case 'number':
+      return {
+        type: 'number',
+        format: String(value),
+      };
+    case 'string':
+      return {
+        type: 'string',
+        format: value === '' ? 'empty' : `"${value}"`,
+      };
+    default:
+      throw new RequestError('swagger.invalid_zod_type', zodLiteral);
```

Comment:
same here should ew just return the value by default?

Reply:
See https://github.com/logto-io/logto/pull/1126#discussion_r898795267